### PR TITLE
Fix useTokenTracker useEffect bug

### PR DESCRIPTION
**Pull request by danjm**
On develop, if selecting a token on the main asset list to go to the token asset view triggers an infinite loop of renders. This seems to be because the last `useEffect` hook inside the custom `useTokenTracker` hook will run whenever the `tokens` object changes, even if it does not _deeply_ change. This ultimately leads to a new `tokensWithBalances` object being set and returned.

This PR corrects this by ensuring that the token `useEffect` hook does not unnecessarily fire when receiving tokens objects that are deeply equal to the previous tokens object received.

This is a naive solution that can probably be made more elegant by those with more hooks expertise than me.

As a demo of the problem, I console.logged the tokens from inside the useEffect hook:
tokens-bug-unfixed

But with this fix applied:
tokens-bug-fixed

### CODE REVIEW

File: ui/app/hooks/useTokenTracker.js [hunk 1]
```diff
@@ -56,6 +65,10 @@ export function useTokenTracker (tokens) {
     return teardownTracker
   }, [teardownTracker])
 
+  const previousTokens = usePrevious(tokens)
+  const previousUserAddress = usePrevious(userAddress)
```
Comment (brad-decker) [inline ui/app/hooks/useTokenTracker.js:69]: you can pass isEqual as a second argument to useSelector to have it do deep comparison and return the same value if inputs are deeply equal... though we've only had to do this in the case where the values are serialized from another value meaning they are always different objects... if there is referential equality then useSelector doesn't return a new object.

tl;dr - I don't think you need to use the usePrevious hook for the values you're pulling from redux .. if you do then just pass isEqual as a second arg to useSelector.

File: ui/app/hooks/useTokenTracker.js [hunk 2]
```diff
@@ -1,8 +1,17 @@
 import { useState, useEffect, useRef, useCallback } from 'react'
 import TokenTracker from '@metamask/eth-token-tracker'
 import { useSelector } from 'react-redux'
+import { isEqual } from 'lodash'
 import { getCurrentNetwork, getSelectedAddress } from '../selectors'
 
+const usePrevious = (value) => {
+  const ref = useRef()
+  useEffect(() => {
```
Comment (brad-decker) [inline ui/app/hooks/useTokenTracker.js:9]: I could be horribly wrong here, but I believe the order of effects matters. These effects will happen before the effect that is checking their value. For this, I think that means they will *always* be equal because the current value is set before you're checking it on line 80. The result would be fixing the infinite renders bug but potentially breaking the effect of reloading tokens when the values change? 

I think what we want is to always return a consistent value for use in the dependency array of the effect in question, similar to passing isEqual to the second argument of useSelector. 

https://github.com/reduxjs/react-redux/blob/a9dce030742385285a243b8769f49c94c5942838/src/hooks/useSelector.js#L30-L74

```js
function useMemoizedValue(value, equalityFn = isEqual) {
  const [, forceRender] = useReducer(s => s + 1, 0)
  const latestValue = useRef();

  useIsomorphicLayoutEffect(() => {
	if (!isEqual(value, latestValue.current)) {
      latestValue.current = value;
      forceRender(); // refs don't cause renders
    }
  }, [value]);

  return latestValue.current;
}
```

downstream usage:

```js
const memoizedTokens = useMemoizedValue(tokens);
```

Disclaimer: I did not test this, but I feel like there's some significant potential here as a tool to prevent unnecessary renders that can potentially impact user experience. We use the second argument of useSelector in other parts of the code base with great effect. CC'ing @Gudahtt here to get a second opinion.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: ui/app/hooks/useEqualityCheck.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * based on an object or value that always changes references but
+ * infrequently changes it's value. By default, uses isEqual from
+ * lodash. This is typically only useful with objects and arrays.
+ *
+ * @param {T}                 value      - any value to check equality of
+ * @param {(T, T) => boolean} equalityFn - A function to determine equality
+ * @returns {T}
+ */
+export function useEqualityCheck (value, equalityFn = isEqual) {
+  const [computedValue, setComputedValue] = useState(() => value)
```
Comment (Gudahtt) [inline ui/app/hooks/useEqualityCheck.js:18]: Why would the default value be a function :thinking: Was this what was intended?
```suggestion
  const [computedValue, setComputedValue] = useState(value)
```

**Review by Gudahtt — COMMENTED** [review]
Great catch! I had known about this issue at one point, and had addressed it in the parent component: https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/components/app/token-list/token-list.js#L18

But it looks like I forgot to address it in the other places this was used :facepalm: Better to fix it here anyway, so this component isn't so easy to misuse.

I just had one minor suggestion - everything looks fine aside from that.

File: ui/app/hooks/useEqualityCheck.js (hunk 3 above)
Comment (brad-decker) [reply ui/app/hooks/useEqualityCheck.js:18]: Hah, I think I got lazy initialization happy with some of Dan's Swaps prs because the default values were being set by iterating through an array.

e.g
```js
const [state, setState] = useState(arrayFromProps.find(i => i.key === keyFromProps))
```

which would do that array iteration every time. So using the lazy evaluation by passing a function means the function only gets called on first render preventing the array loop on every render. It's definitely not saving anything here except allocating an additional function every render needlessly. Thanks for catching that.

**Comment by danjm** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by brad-decker — APPROVED** [review]
